# Week 2 Checkpoint: First Data Profile
**BAN 6003 Data Management and Analytics Integration**

Complete this notebook using the `employee_data.csv` file in the `data/` folder.

This is the independent checkpoint for Week 2. Use the functions practiced in the guided lab. You are not expected to clean the data yet. Your goal is to profile the data and report what you notice.


## Instructions

1. Run each cell from top to bottom.
2. Fill in all cells marked `# Your code here`.
3. Write short answers in the markdown response areas.
4. Save, commit, and push your notebook to GitHub.

You may use the Week 2 guided lab as a reference.


## Task 0: Import Pandas and Load the Data
Run the cell below.


In [16]:
import pandas as pd

df = pd.read_csv("../data/employee_data.csv")
df.head()


,employee_id,first_name,last_name,hire_date,department,salary,age,full_time
0,101,John,Smith,4/15/22,Marketing,85000,34.0,True
1,102,Jane,Doe,8/20/21,Sales,92000,45.0,True
2,103,Peter,Jones,1/2/22,HR,71000,29.0,True
3,104,Mary,Johnson,11/10/23,Engineering,115000,38.0,True
4,105,Mike,Brown,5/30/20,sales,89000,NaN,False


## Task 1: Basic Structure
Show the basic structure of the DataFrame.

Your code should show:
- the shape of the DataFrame;
- the column names;
- the first five rows;
- the last five rows.


In [18]:
# Your code here
print("Shape of DataFrame")
print(df.shape)

print("Column names")
print(df.columns)

print("First 5 rows and Last 5 rows")
pd.concat([df.head(), df.tail()])

Shape of DataFrame
(13, 8)
Column names
Index(['employee_id', 'first_name', 'last_name', 'hire_date', 'department',
       'salary', 'age', 'full_time'],
      dtype='str')
First 5 rows and Last 5 rows


,employee_id,first_name,last_name,hire_date,department,salary,age,full_time
0,101,John,Smith,4/15/22,Marketing,85000,34.0,True
1,102,Jane,Doe,8/20/21,Sales,92000,45.0,True
2,103,Peter,Jones,1/2/22,HR,71000,29.0,True
3,104,Mary,Johnson,11/10/23,Engineering,115000,38.0,True
4,105,Mike,Brown,5/30/20,sales,89000,NaN,False
8,102,Emily,White,1/15/24,Sales,91000,31.0,True
9,110,David,Miller,7/22/19,HR,68000,58.0,True
10,111,Sarah,Moore,3/12/23,Marketing,"$91,000",28.0,True
11,error,Kevin,Clark,2/5/24,Sales,88000,35.0,True
12,113,Brian,Hall,6/18/22,Engineering,5900000,999.0,True


### Task 1 Written Response
Write 1–2 sentences describing the size and general structure of the dataset.

**Your response:**  
The dataframe has 13 rows and 8 columns.  
All columns are string data types.

## Task 2: Data Types and Non-Missing Counts
Use Pandas to inspect data types and non-missing counts.

Your code should use:
- `info()`
- `dtypes`


In [24]:
# Your code here
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 13 entries, 0 to 12
Data columns (total 8 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   employee_id  13 non-null     str    
 1   first_name   13 non-null     str    
 2   last_name    13 non-null     str    
 3   hire_date    12 non-null     str    
 4   department   13 non-null     str    
 5   salary       12 non-null     str    
 6   age          12 non-null     float64
 7   full_time    12 non-null     object 
dtypes: float64(1), object(1), str(6)
memory usage: 964.0+ bytes


### Task 2 Written Response
Which columns, if any, have data types that might need to be investigated later? For example, is a number or date stored as text/object?

**Your response:**  
Hire date, salary, age, and full time columns contain missing data.
Salary should be numeric, but is listed as a string.
Hire date should be a date, but is listed as a string.


## Task 3: Missing Values
Check missing values for all columns.

Your code should use `isna().sum()`.

Optional: also calculate missing percentages.


In [33]:
# Your code here
print("Total of Missing Data")
print(df.isna().sum())

print("Percent of Data Missing")
print(df.isna().mean())

Total of Missing Data
employee_id    0
first_name     0
last_name      0
hire_date      1
department     0
salary         1
age            1
full_time      1
dtype: int64
Percent of Data Missing
employee_id    0.000000
first_name     0.000000
last_name      0.000000
hire_date      0.076923
department     0.000000
salary         0.076923
age            0.076923
full_time      0.076923
dtype: float64


### Task 3 Written Response
Which columns have missing values? Which missing values seem most important to investigate later?

**Your response:**  
Hire date, salary, age, and full time columns contain missing data.


## Task 4: Numerical Summary
Use `.describe()` to summarize numerical columns.

Then choose one numerical column and calculate at least two individual summary statistics such as mean, median, min, max, or standard deviation.


In [38]:
# Your code here
print(df.describe())

print("Lowest Age:", df["age"].min())
print("Highest Age:", df["age"].max())


              age
count   12.000000
mean   118.583333
std    277.411889
min     28.000000
25%     32.500000
50%     36.500000
75%     46.750000
max    999.000000
Lowest Age: 28.0
Highest Age: 999.0


### Task 4 Written Response
What do you notice from the numerical summaries? Mention at least one range, unusual value, or pattern.

**Your response:**  
The range of age goes up to 999.  This is skewing the calculations for mean and standard deviation.  Must investigate why this value is so high.


## Task 5: Categorical Summary
Use `.value_counts()` to summarize at least one categorical column.

Start with `department` if it exists in the dataset. Then choose one additional categorical column if available.


In [42]:
# Your code here
print(df["department"].value_counts(dropna=False))

print(df["employee_id"].value_counts(dropna=False))

department
Sales          4
Marketing      3
Engineering    3
HR             2
sales          1
Name: count, dtype: int64
employee_id
102      2
101      1
103      1
104      1
105      1
106      1
107      1
108      1
110      1
111      1
error    1
113      1
Name: count, dtype: int64


### Task 5 Written Response
What are the most common categories? Do you notice anything that might need standardization or further investigation later?

**Your response:**  
The most common department is Sales; however, department names are not standardized.  For example, 'Sales' and 'sales' are listed as separate departments.
For employee ids, there are two listings for 102, which could indicated an possible incorrect duplication; there is also an id listed as 'error'.


## Task 6: Duplicate Checks
Check for duplicate records and duplicate employee IDs.

Your code should check:
- fully duplicated rows;
- duplicate `employee_id` values;
- the most common `employee_id` values using `value_counts()`.


In [47]:
# Your code here
print("# of Duplicate rows:", df.duplicated().sum())

print("# of Duplicate Employee IDs:", df["employee_id"].duplicated().sum())

print(df["employee_id"].value_counts())

# of Duplicate rows: 0
# of Duplicate Employee IDs: 1
employee_id
102      2
101      1
103      1
104      1
105      1
106      1
107      1
108      1
110      1
111      1
error    1
113      1
Name: count, dtype: int64


### Task 6 Written Response
Are there duplicate employee IDs? If yes, what might that mean if `employee_id` is supposed to uniquely identify each employee?

**Your response:**  
Employee ID 102 is listed twice.  This could be a data entry error, or the same employee is list twice for some reason.  For example: not updating the employee file correctly (new entry for promotion, name change, etc).  


## Task 7: Overall Data Quality Observations
Write 3–5 bullet points summarizing the condition of the dataset.

Focus on what you found from profiling. Do **not** clean the data yet.

Examples:
- The dataset has 13 rows and 8 columns.
- Hire date, salary, age, and full time columns contain missing data.
- Salary should be numeric, but is listed as a string.
- Hire date should be a date, but is listed as a string.
- The Employee ID column has one duplicate ID and one listed as 'error'.
- The age column has a range from 28 to 999.


**Your observations:**

1. The dataset has 13 rows and 8 columns.
2. Hire date, salary, age, and full time columns contain missing data.
3. Salary should be numeric, but is listed as a string.
4. Hire date should be a date, but is listed as a string.
5. The Employee ID column has one duplicate ID and one listed as 'error'.
6. The age column has a range from 28 to 999.


## Submission Checklist
Before submitting, make sure:

- [ ] All code cells run without errors.
- [ ] You completed all tasks.
- [ ] You wrote short responses in the markdown cells.
- [ ] You saved the notebook.
- [ ] You committed and pushed your work to GitHub.

Use the terminal:

```bash
git add .
git commit -m "Completed Week 2 data profile checkpoint"
git push
```
